src/libexpr/primops/fromYAML.cc

```diff
@@ -0,0 +1,156 @@
+#ifdef HAVE_RYML
+
+#include "primops.hh"
+#include "eval-inline.hh"
+
+#include <ryml.hpp>
+#include <c4/format.hpp>
+#include <c4/std/string.hpp>
+
+
+namespace nix {
+
+struct NixContext {
+    EvalState & state;
+    const PosIdx pos;
+    std::string_view yaml;
+};
+
+static void s_error [[ noreturn ]] (const char* msg, size_t len, ryml::Location, void *nixContext)
+{
+    auto context = static_cast<const NixContext *>(nixContext);
+    if (nixContext) {
+        throw EvalError({
+            .msg = hintfmt("while parsing the YAML string '%1%':\n\n%2%",
+                context->yaml, std::string_view(msg, len)),
+            .errPos = context->state.positions[context->pos]
+        });
+    } else {
+        throw EvalError({
+            .msg = hintfmt("failed assertion in rapidyaml library:\n\n%1%",
+                std::string_view(msg, len))
+        });
+    }
+}
+
+static void visitYAMLNode(NixContext & context, Value & v, ryml::ConstNodeRef t) {
+
+    auto valTypeCheck = [=] (ryml::YamlTag_e tag, bool defaultVal = true) {
+        auto valTag = ryml::TAG_NONE;
+        if (t.has_val_tag()) {
+            auto tag = t.val_tag();
+            valTag = tag == "!" ? ryml::TAG_STR : ryml::to_tag(tag);
+        }
+        return valTag == ryml::TAG_NONE ? defaultVal : valTag == tag;
+    };
+
+    v.mkBlackhole();
+    if (t.has_key_tag()) {
+        auto tag = ryml::to_tag(t.key_tag());
+        if (tag != ryml::TAG_NONE && tag != ryml::TAG_STR) {
+            auto msg = ryml::formatrs<std::string>(
+                "Error: Nix supports string keys only, but the key '{}' has the tag '{}'", t.key(), t.key_tag());
+            s_error(msg.data(), msg.size(), {}, &context);
+        }
+    }
+    if (t.is_map()) {
+        if (t.num_children() == 1) { // special case for YAML string ":"
+            auto child = t.child(0);
+            if (child.key().empty() && !child.is_key_quoted() && child.has_val() && child.val().empty() && !child.is_val_quoted()) {
+                v.mkNull();
```

Comment:
What does 'special case for YAML string ":"' mean? Why does it map to null (rather than, say, an error)?

Reply:
Good catch. Actually this should throw an error. Now I recognize that I'm parsing the test files incorrectly. YAML `:` is equivalent to YAML `null: null` (see [Example 8.18](https://yaml.org/spec/1.2.2/#822-block-mappings)), but this cannot be converted to JSON, i.e. it's not equivalent to JSON `{ "null" :null}` or `null`.